### PR TITLE
http artwork / cover support

### PR DESCRIPTION
**Pull request by whatdoineed2do**
Functionality to support web ui requesting artwork for track/album/artist (requests https://github.com/ejurgensen/forked-daapd/issues/563 and https://github.com/chme/forked-daapd-web/issues/5) - primary use to display cover art on `now playing` screen.

* Added `httpd_artwork` to serve png/jpeg artwork (if available) from current caching/artwork mechanism as used by mpd / roap.  
* Updated `httpd_jsonapi:jsonapi_reply_player()` to add `file_id` to response to allow web client to determine if artwork available for current track

Relies on existing artwork location/handling and caching structures in the backend (no adds for this feature)

The `jsonapi_reply_player()` does not send a full URI to the client (it sends `file_id` to be used to ask server for `/artwork/track/<file_id>`) to avoid complications of multi-homed server and needing to determine what ip/addr the webclient thinks the server is on (didn't find the `accept`d socket in `httpd_request` to call `getpeername()` to build the required URI)

Frontend (pending PR to forked-daapd-web)
* https://github.com/whatdoineed2do/forked-daapd-web/tree/httpd_artwork
* built/drop-in `htdocs` file: [dist-httpd-artwork.tar.gz](https://github.com/chme/forked-daapd-web/files/2320552/dist-httpd-artwork.tar.gz)

The `now playing` screen displays artwork (in square format) ONLY if artwork available above the playhead scrubber and below the existing info (track/album/artist).  With no artwork the screen looks like it does now

screenshot at 2018-08-26 10-23-58
screenshot at 2018-08-26 10-26-35

### CODE REVIEW

**Comment by ejurgensen** [comment]
Sorry about taking a while to look at this, had to prioritize other issues. I took an initial look now, but before I continue I must ask you to fix the pr so it follows the same coding standards as the rest of forked-daapd. If everyone uses their own style the project will become a mess.

**Comment by whatdoineed2do** [comment]
coding style - years of habit!  Fixed - it should only be in `httpd.c`, the new `httpd_artwork.c` files should fit existing style.

File: src/httpd_artwork.c [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+  struct db_group_info dbgri;
+  struct query_params query_params;
+  int ret;
+
+  if (artist_id == NULL && album_id == NULL || (artist_id && album_id)) 
+    {
+      return -1;
+    }
+  memset(&query_params, 0, sizeof(struct query_params));
+  memset(&dbgri, 0, sizeof(dbgri));
```
Comment (ejurgensen) [inline src/httpd_artwork.c]: Use sizeof(struct ...), it protects against errors if you or someone else one day changes dbgri to a pointer.

File: src/httpd_artwork.c [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+      snprintf(clen, sizeof(clen), "%ld", (long)len);
+      evhttp_add_header(headers, "Content-Length", clen);
+
+      ret = 0;
+    }
+  return ret;
+}
+
+static int
+fetch_artwork(struct httpd_request *hreq, const char *artist_id, const char *album_id)
```
Comment (ejurgensen) [inline src/httpd_artwork.c]: What is the purpose of the db lookup? The artwork module will do that for you, so why not just yank the id's straight into artwork_get_xxxx?

File: src/httpd_artwork.c [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+  if ( (dbgri == NULL && dbmfi_id == NULL) || (dbgri && dbmfi_id)) 
+    {
+      return -1;
+    }
+
+  if (dbgri)
+    {
+      int  gpid;
+      safe_atoi32(dbgri->id, &gpid);
+      if ( (ret = artwork_get_group(hreq->reply, gpid, 600, 600)) < 0)
```
Comment (ejurgensen) [inline src/httpd_artwork.c]: Default dimensions are fine, but there should be params for requesting other sizes

File: src/httpd_artwork.c (hunk 2 above)
Comment (whatdoineed2do) [reply src/httpd_artwork.c]: The web client could ask for artwork for a given `artist` or `album`;  the artwork i/f provides `artwork_get_group()` which requires `group id` but I didn't see anything that could transform a `artist_id` / `album_id` to `group id` for that call hence the db lookups here.

`artwork_get_item()` takes the `media file info` id which again does not appear to pull for the group.

File: src/httpd_artwork.c [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+
+  CHECK_NULL(L_WEB, hreq->reply = evbuffer_new());
+
+  status_code = hreq->handler(hreq);
+
+  switch (status_code)
+    {
+      case HTTP_OK:                  /* 200 OK */
+	headers = evhttp_request_get_output_headers(req);
+        if ( (evhttp_find_header(headers, "Content-Type")) == NULL) {
```
[the remark this answers is not in the document]
Comment (whatdoineed2do) [reply src/httpd_artwork.c]: this is redundant case .. removed

File: src/httpd_artwork.c (hunk 2 above)
Comment (ejurgensen) [reply src/httpd_artwork.c]: Yes, I was thinking the client would use the group id. Is it not available to the client? If not, perhaps it could be?

**Comment by chme** [comment]
Instead of adding a `file_id` attribute to the player state in the JSON API and require the clients to construct the artwork url, imo it is better to add an `artwork_url` attribute to the queue item object. 

Artwork is part of the metadata of an album, track and queue item. It is always displayed in that context and therefor in the JSON API it should be part of the album, track or queue_item object and not the player state object.

I also do not think, that it is good to require the clients to know how to construct an artwork url. Instead the url should be constructed in httpd_jsonapi.c and returned to the client as an "artwork_url" attribute. In the future, this will allow to support artwork for items that are not in the library (e. g. radio streams that are not in a playlist or spotify tracks that are not in the user library). These non persistent items all have the same file_id = 9999999.

**Comment by ejurgensen** [comment]
Yes, sounds sensible

**Comment by whatdoineed2do** [comment]
`artwork_url` as part of `queue` is good for me.  But to confirm, what are we expecting to be constructed and in the queue item?  The full http:// address or what the web ui is constructing and sending to the server (`/api/artwork/track/<file_id>` so they'd still have to hack on `http://<hostname:port>/)?

One reason I went with getting the client to construct the address was because i wasnt sure where I'd get the server address to ask the client to hit (multihomed hosts etc)

**Comment by whatdoineed2do** [comment]
Made some mods to this PR so that the (base) artwork url (ie `/artwork/track/<id>` where `id` is the `file_id` to allow looking via the `artwork.c` interfaces) is stored in the `queue_items` but only for local files.

Upon the web interface asking for details for the `now playing` screen the artwork url is sent back with fully qualified with the server's address as seen by the client (this should take care of ppl running this in containers or behind reverse proxies etc).

At the web player side, it will be presented with a URL it can hit for artwork.  If no artwork is available, empty artwork URL is sent.

Let me know if this is more inline with your thoughts above.

**Comment by ejurgensen** [comment]
This pr is still not quite there. I will try and fix it up myself and then merge it. Will probably be within a week or two.

**Comment by whatdoineed2do** [comment]
Thanks, appreciated.  I'll pull the webui merge from master and send through the front end changes to support artwork so this can all go out together.

**Comment by ejurgensen** [comment]
You can see from the branch https://github.com/ejurgensen/forked-daapd/tree/whatdoineed2do-httpd_artwork how I am planning to implement it. When I merge it, I will squash all the commits in the branch.

The artwork url's will be relative. There will be no guarantee that there actually is any artwork, because checking for that would make the json api much slower + doesn't really fit with how the artwork module works.

**Comment by whatdoineed2do** [comment]
Your impl is cleaner.  One Q though:

> The artwork url's will be relative

`snprintf(artwork_url, sizeof(artwork_url), "/artwork/item/%s", dbmfi->id);`

Using relative paths would mean client side needs to construct the full URI before coming back for the artwork which I thought was decided against?  Is it worth constructing the full URI as we have the `evhttp_request` object..
```
jsonapi_artwork_uri(struct evhttp_request *ev_req, char* artwork)
{
...
  /* figure out where the cli request thinks this server is located via 
   * http/1.1 mandatory 'host' header
   */
  if ( (host = evhttp_find_header(evhttp_request_get_input_headers(ev_req), "Host")) == NULL)
    {
      DPRINTF(E_DBG, L_WEB, "missing 'Host' http request header\n");
      return NULL;
    }
  len = 8 + strlen(host) + strlen(artwork) +1;
  buf = (char*)malloc(len);
  snprintf(buf, len, "%s://%s%s", 
              ( (proto = evhttp_find_header(evhttp_request_get_input_headers(ev_req), "X-Forwarded-Proto")) == NULL ?
                    "http" : proto),
              host,
              artwork);
  DPRINTF(E_DBG, L_WEB, "full artwork uri=%s\n", buf);
  
  return buf;
}
```

**Comment by ejurgensen** [comment]
I have merged the refactored version now

**Pull request closed without merging**